doc/sphinx/proofs/writing-proofs/rewriting.rst

```diff
@@ -1,102 +1,120 @@
+   .. todo PR: the definition in .v file is a little different
 
-   This tactic applies to any goal. The type of :token:`term` must have the form
+- :gdef:`Setoid equality <setoid equality>` defines equality in terms of an equivalence
+  relation.  A :gdef:`setoid` is a set that is equipped with an equivalence relation
+  (see https://en.wikipedia.org/wiki/Setoid).  These are needed to form a :gdef:`quotient set`
+  or :gdef:`quotient`
+  (see https://en.wikipedia.org/wiki/Equivalence_Class).  In Coq, users generally work
+  with setoids rather than constructing quotients, for which there is no specific support
+  (see also `#10871 <https://github.com/coq/coq/issues/10871>`_).
 
-   ``forall (x``:sub:`1` ``:A``:sub:`1` ``) ... (x``:sub:`n` ``:A``:sub:`n` ``), eq term``:sub:`1` ``term``:sub:`2` ``.``
+- :gdef:`Definitional equality <definitional equality>` is equality based on the conversion rules, which
+  Coq can determine automatically.  When two terms are definitionally equal, Coq knows it can
+  replace one with the other, such as with :tacn:`change` `X with Y`, among many
+  other advantages.
 
-   where :g:`eq` is the Leibniz equality or a registered setoid equality.
+  .. todo PR: A and B are convertible iff they are definitionally equal??
 
-   Then :n:`rewrite @term` finds the first subterm matching `term`\ :sub:`1` in the goal,
-   resulting in instances `term`:sub:`1`' and `term`:sub:`2`' and then
-   replaces every occurrence of `term`:subscript:`1`' by `term`:subscript:`2`'.
-   Hence, some of the variables :g:`x`\ :sub:`i` are solved by unification,
-   and some of the types :g:`A`\ :sub:`1`:g:`, ..., A`\ :sub:`n` become new
-   subgoals.
+.. _rewritingexpressions:
 
-   .. exn:: The @term provided does not end with an equation.
-      :undocumented:
+Rewriting expressions
+---------------------
 
-   .. exn:: Tactic generated a subgoal identical to the original goal. This happens if @term does not occur in the goal.
-      :undocumented:
+These tactics use the equality :g:`eq : forall {A : Type}, A -> A -> Prop` defined in
+file ``Logic.v`` (see :ref:`coq-library-logic`). The notation for :g:`eq t u` is
+:g:`t = u`.
 
-   .. tacv:: rewrite -> @term
+.. tacn:: rewrite {+, @oriented_rewriter } {? @occurrences } {? by @ltac_expr3 }
 
-      Is equivalent to :n:`rewrite @term`
+   .. insertprodn oriented_rewriter one_term_with_bindings
 
-   .. tacv:: rewrite <- @term
+   .. prodn::
+      oriented_rewriter ::= {? {| -> | <- } } {? @natural } {? {| ? | ! } } @one_term_with_bindings
+      one_term_with_bindings ::= {? > } @one_term {? with @bindings }
 
-      Uses the equality :n:`@term`:sub:`1` :n:`= @term` :sub:`2` from right to left
+.. todo PR: This description is wrong.  This doesn't work:
+   Goal forall x: nat, x = 2 -> x = 0 /\ x = 1.
+   intro. intro H.
+   rewrite (forall (x:nat), eq x 3).
+   But rewrite H. does.
+   Also want to check if finds subterms inside forall, thought I had an example where it didn't.
```

Comment:
You are correct. It is wrong to say that `one_term` must have the following form. It is its type that must have this form. Thus, typically this `one_term` will be a hypothesis or theorem name.

The answer to your second question is that `rewrite` cannot find subterms inside forall, but `setoid_rewrite` can.

Reply:
I suppose everything shown in the context is a type, but it seems to me referring to `type of` here isn't very clear because we don't generally use that language.  We just refer to a hypothesis.  Does that make sense?

```
1 subgoal
x : nat
H : x = 2
______________________________________(1/1)
x = 0 /\ x = 1
```

Reply:
Everything after a colon is a type. `H` is the name of a hypothesis but also a (very simple) term. Its type is `x = 2`. We tend to confuse the hypothesis term / name (`H`) and type (`x = 2`) indeed!

Reply:
I'd like to add a couple paragraphs somewhere that show the current goal and explain the terms "hypothesis" "conclusion" "current goal" "shelved goal" etc. and mention that the part after the : and below the line is actually a type.  Where do you suggest?  In Proof Handling?

Reply:
Yes, absolutely! Note that I've recently renamed this file into `proof-mode.rst` and I was hoping to change the title as well. What you propose would be a very nice introduction to this chapter.

Reply:
> The answer to your second question is that rewrite cannot find subterms inside forall, but setoid_rewrite can.

This isn't quite accurate; rewrite will find all and only the subterms of a forall that don't mention the forall binders.  So with `x = y`, you can still rewrite `forall a, x = a`.

Reply:
Thanks.  I have this now:

```[rewrite] won't find occurrences inside `forall` that refer to variables bound by the `forall` ```